docs/docs/protocol-specs/gas-and-fees/fee-schedule.md

```diff
@@ -1,26 +1,100 @@
-| Verifying each nullifier against the world state    | L2     | Fixed L2/Tx nullifier     | |
-| Verifying each nullifier against others in the same block     | L2     | Fixed L2/Tx nullifier     | Whilst not strictly a fixed cost, this would need to be allocated as a fixed cost as it depends on the composition of the rollup |
-| Verifying log preimages against the sha256 log hashes contained in the private kernel public inputs | L2 | L2 gas per pre-image field | |
-| Verifying contract deployment data against the sha256 hash of this data contained in the private kernel public inputs | L2 | L2 gas per hash | |  
-| Publishing contract data to DA     | DA     | DA gas per byte     | |
-| Publishing state updates to DA     | DA     | DA gas per byte     | |
-| Publishing notes/tags to DA    | DA     | DA gas per byte     | |
-| Publishing L2->L1 messages | L1 | Calldata gas per byte + processing & storing of the message | |
-| Public function execution    | L2     | L2 gas per function opcode     | |
-| Proving the public VM circuit for a public function     | L2     | Fixed L2/Tx public function     | |
-| Proving the public kernel circuit for a public function    | L2     | Fixed L2/Tx public function   | |
-| Proving the base rollup circuit | L2 | Fixed L2/Transaction |
-| Proving the merge rollup circuit | L2 | Amortized L2/Transaction |
-| Proving the root rollup circuit | L2 | Amortized L2/Transaction |
-| Publishing the block header to L1 | L1 | Amortized L1/Transaction |
-| Verifying the rollup proof | L1 | Amortized L1/Transaction |
\ No newline at end of file
+# L2 Gas
+
+L2 gas is consumed to cover the costs associated with executing the public VM, proving the public VM circuit, and proving the public kernel circuit.
+
+The public vm has an [instruction set](../public-vm/instruction-set.mdx) with opcode level gas metering to cover the cost of actions performed within the public VM.
+
+Additionally, there is a fixed cost associated with each iteration of the public VM (i.e. the number of enqueued public function calls, plus 1 if there is a teardown function), which is used to cover the cost of proving the public VM circuit.
+
+The L2 gas used is then calculated as:
+
+```
+AVM_STARTUP_L2_GAS = 1024
+
+
+num_avm_invocations = (number of enqueued public function calls) +
+                      (is there a teardown function ? 1 : 0)
+
+l2_gas_used = AVM_STARTUP_L2_GAS * num_avm_invocations + 
+                teardown_l2_gas +
+                (gas reported as consumed by the public VM)
+```
+
+# Max Inclusion Fee
+
+Each transaction, and each block, has inescapable overhead costs associated with it which are not directly related to the amount of data or computation performed. 
+
+These costs include:
+- verifying the private kernel proof of each transaction
+- executing/proving the base/merge/root rollup circuits
+  - includes verifying that every new nullifier is unique across the tx/block
```

Comment:
Agreed! I think that this will be baked into the L2 gas cost of the `NULLIFIEREXISTS` AVM op code. Similar to:
- `NOTEHASHEXISTS`
- `EMITNOTEHASH`
- `EMITNULLIFIER`

Reply:
But isn't that only for nullifiers emitted during public calls?

Reply:
I added a note to this point above.